### PR TITLE
fix(cli): build assets before deploying any stacks

### LINKED ISSUE #21511: (cli): build+publish all assets used for all selected stacks before deploying any stacks
### Describe the feature

The AWS CDK should support building and publishing all relevant assets from all selected stacks before it deploys anything.

### Use Case

When deploying more than one stack at a time, and when these stacks contain docker image assets, `docker build` can happen after you've already deployed a few stacks. But, `docker build` can fail for many reasons, including unexpected rate limiting. When this happens, the CDK will deploy some stacks but not all. The CDK could prevent this failure state if it built and published all relevant assets *before* it deployed anything with CloudFormation.

### Proposed Solution

Publishing could be performed before https://github.com/aws/aws-cdk/blob/220177fdb7aecc2764ffca31c48004fd54275e3a/packages/aws-cdk/lib/cdk-toolkit.ts#L169.

Most of the code needed is already in `CloudFormationDeployments.publishStackAssets`:
https://github.com/aws/aws-cdk/blob/74318c7d22bfc00de9e005f68a0a6aaa58c7db39/packages/aws-cdk/lib/api/cloudformation-deployments.ts#L452

### Other Information

I've opened a PR with the relevant changes to enable this behaviour. #21513

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.35.0

### Environment details (OS name and version, etc.)

Linux

### PR DESCRIPTION
**Pull request by misterjoshua**
Changes the CDK CLI to build assets before deploying any stacks. This allows the CDK CLI to catch docker build errors, such as from rate limiting, before any stacks are deployed. Moving asset builds this early prevents these build failures from interrupting multi-stack deployments part way through. 

Fixes #21511

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by misterjoshua** [comment]
Here's a demonstration of how this looks on a local test project:

https://user-images.githubusercontent.com/644092/183525916-18647ec3-d273-43a6-97ef-7b1642249fa0.mp4

**Comment by misterjoshua** [comment]
I'm unsure whether I should be updating a README for this one - should I rename this to a `fix` so the PR linter doesn't complain?

**Comment by niebloomj** [comment]
Would this update Lambda code also before deploying stacks?

**Comment by misterjoshua** [comment]
> Would this update Lambda code also before deploying stacks?

@niebloomj Nope. It stages Lambda's code in a bucket or ECR repository, but it continues to wait for the deploy stage to come around before any changes occur. Only when deploy comes around the Lambda function is updated to point to the newly uploaded asset.

File: packages/aws-cdk/lib/api/cloudformation-deployments.ts [hunk 1]
```diff
@@ -338,9 +345,14 @@ export class CloudFormationDeployments {
...
-      await this.publishStackAssets(options.stack, toolkitInfo);
+    const disableAssetPublishing = options.disableAssetPublishing ?? false;
+    const isImporting = options.resourcesToImport !== undefined;
+    // Determine whether we should publish assets. We don't publish if assets
+    // are prepublished or during an import operation.
+    const shouldPublish = !disableAssetPublishing && !isImporting;
+
+    if (shouldPublish) {
+      await this._publishStackAssets(options.stack, toolkitInfo);
     }
```
Comment (misterjoshua) [inline packages/aws-cdk/lib/api/cloudformation-deployments.ts:356]: All this is to maintain `deployStack()`'s backwards compatibility, but I wonder whether publishing assets via `deployStack()` is even necessary. Given that `cdk import` is the only other place `deployStack()` is used and that we're not publishing assets there, I can't think of any good reasons to keep the old behaviour. But, I wanted to discuss this first before making that change.

Comment (mrgrain) [reply packages/aws-cdk/lib/api/cloudformation-deployments.ts:356]: `CloudFormationDeployments` is not exposed in v2 (though it was in v1). See https://github.com/aws/aws-cdk/issues/18211

I'd want someone else to confirm this, but from what I can tell it's up for changes.

**Comment by mrgrain** [comment]
@misterjoshua I'm a concerned this PR reverts recent gains made by #19378
Specifically it looks like asset publishing is now a synchronous operation for all stacks.

**Comment by misterjoshua** [comment]
> @misterjoshua I'm a concerned this PR reverts recent gains made by #19378 Specifically it looks like asset publishing is now a synchronous operation for all stacks.

@mrgrain It should be possible to parallelize this build process at the given level of concurrency. I'll see what I can do about that.

**Comment by mrgrain** [comment]
Does this respect `--exclusively` ?

**Comment by misterjoshua** [comment]
> Does this respect `--exclusively` ?

@mrgrain Yes, it does. The stack collection we're drawing from is provided by `CdkTooklit.selectStacksForDeploy` which handles the `--exclusively` flag for us. When `--exclusively` is requested, the stack collection will contain just one stack.

https://github.com/aws/aws-cdk/blob/c81068b34f62a2daef306840cfb44d332d8fe6cf/packages/aws-cdk/lib/cdk-toolkit.ts#L141
https://github.com/aws/aws-cdk/blob/c81068b34f62a2daef306840cfb44d332d8fe6cf/packages/aws-cdk/lib/cdk-toolkit.ts#L630

**Comment by misterjoshua** [comment]
> @misterjoshua I'm a concerned this PR reverts recent gains made by #19378 Specifically it looks like asset publishing is now a synchronous operation for all stacks.

@mrgrain I've added concurrency to the publishing step. Does this alleviate your concern?

**Comment by mrgrain** [comment]
> @mrgrain I've added concurrency to the publishing step. Does this alleviate your concern?

Yes, it does address the speed improvements.

--- 

In the PR and issue you seem to imply that your motivation of always failing ALL stack deployments when there is an issue with assets in ANY of the stacks is a desirable or even expected behavior. I can see this very much being the case for App with coupled stacks. The current functionality however is that any stacks before the failure will deploy successfully. Admittedly `--concurrency`  is brand-new, but with this flag I believe _all_ other stacks would deploy successfully.

This begs the question to me: Can the change in this PR be the unchangeable default behavior? Should it be an optional flag or even pulled out into separate command?

**Review by mrgrain — CHANGES_REQUESTED** [review]
Apologies forgot to post these pending comments.

File: packages/aws-cdk/lib/cdk-toolkit.ts [hunk 2]
```diff
@@ -169,6 +169,23 @@ export class CdkToolkit {
...
 
+    // Asset pre-publishing phase.
+    for (const stack of stacks) {
+      // Check whether the stack has an asset manifest before trying to
+      // pre-build and publish.
+      if (!stack.dependencies.some(cxapi.AssetManifestArtifact.isAssetManifestArtifact)) {
+        continue;
+      }
+
+      print('🏭 Pre-building and publishing assets for %s\n', stack.displayName);
```
[the remark this answers is not in the document]
Comment (mrgrain) [reply packages/aws-cdk/lib/cdk-toolkit.ts]: Re-reading this - Building usually occurs at synth time. This might be confusing.

**Comment by rix0rrr** [comment]
Hi @misterjoshua, thanks for the good idea. I see how this would definitely help with common pain points.

Unfortunately,  this change  is going to interfere with future changes I would like to make. Those being:

* treating asset deployments as "just any other type of artifact that can be orchestrated"; as well as
* giving them inputs and outputs.

Doing so would make it possible to do things like: stack 1 creates a bucket, the assets for stack 2 are published into that bucket, then stack 2 gets deployed -- and it all gets cleaned up on deletion.

I'm pretty sure many people would probably much prefer this (it makes it so that the bootstrap stack doesn't need to pre-create S3 buckets and ECR repositories), but that requires that assets are not "special" and just get sequenced like any other deployment.

Treating assets specially and pulling their publishing to the front would not be compatible with that model.

------

In the hypothetical future situation, we can achieve the current behavior in 2 ways:

* Scan the list of assets, build them first, then proceed with deploying as normal; or
* Separate out the building and publishing of assets as 2 different types of actions orchestrated by the CLI, in which case we can push all builds to the front and sequence the deploys as usual.

I think the second solution would have my preference, but I can live with the first.

----------

With respect to this PR, in the short term you can try one of these:

* Help work towards our utopia maybe? (😅 -- but it will be a large lift) 
* Do the short-circuit solution, in which we pre-build assets (but publish at the normal time)
* Solve the issue locally for yourself, by running `cdk-assets` before running `cdk deploy`. Short-circuiting will make sure no duplicate work is done (However, the `cdk-assets` CLI doesn't have the auth flexibility that the CDK CLI itself has, and depending on your auth setup may or may not work)

**Review by rix0rrr — CHANGES_REQUESTED** [review]
See my previous post

**Comment by misterjoshua** [comment]
> ...
> With respect to this PR, in the short term you can try one of these:
>     * Help work towards our utopia maybe? (😅 -- but it will be a large lift)
>     * Do the short-circuit solution, in which we pre-build assets (but publish at the normal time)
>     * Solve the issue locally for yourself, by running `cdk-assets` before running `cdk deploy`. Short-circuiting will make sure no duplicate work is done (However, the `cdk-assets` CLI doesn't have the auth flexibility that the CDK CLI itself has, and depending on your auth setup may or may not work)

@rix0rrr Thanks for these insights. I like your longer-term plan. I'll start looking at the work to do either of your first two options for this PR, as the third about cdk-assets isn't ideal for us.

**Comment by misterjoshua** [comment]
@mrgrain In light of Rico's recent comments, I'll have to get back to you on your review after I decide with my team which way we're going. Thank you for all your help reviewing so far.

**Comment by misterjoshua** [comment]
@mergifyio update

**Comment by misterjoshua** [comment]
@rix0rrr I've adjusted this PR to separate asset builds from publishing. Is this closer to what you had in mind?

**Comment by misterjoshua** [comment]
I've fixed a few small oversights and recorded another video to show how this looks for three stacks.

https://user-images.githubusercontent.com/644092/184549129-39fe064d-46cf-48c7-9566-bf914a8151aa.mp4

**Comment by mrgrain** [comment]
@misterjoshua That looks like the right direction, but let's wait for @rix0rrr 's feedback first.

**Comment by misterjoshua** [comment]
> @misterjoshua That looks like the right direction, but let's wait for @rix0rrr 's feedback first.

Sounds good. I'll keep an eye out for comments. :)

File: packages/aws-cdk/lib/build.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import * as cxapi from '@aws-cdk/cx-api';
+
+type Options = {
+  buildStackAssets: (stack: cxapi.CloudFormationStackArtifact) => Promise<void>;
+};
+
+export async function buildAllStackAssets(stacks: cxapi.CloudFormationStackArtifact[], options: Options): Promise<void> {
```
Comment (misterjoshua) [inline packages/aws-cdk/lib/build.ts:7]: This function used to handle `--concurrency`, but after #21661 reverted `--concurrency`, we have to build sequentially. As-is, this logic could arguably be moved into `CdkToolkit`. But, I wanted to wait until I get more feedback on whether this PR is going in the right direction before I make too many changes.

File: packages/aws-cdk/lib/api/cloudformation-deployments.ts [hunk 4]
```diff
@@ -236,6 +236,13 @@ export interface DeployStackOptions {
    * @default - Use the stored template
    */
   readonly overrideTemplate?: any;
+
+  /**
+   * Disable asset publishing
+   *
+   * @default false
+   */
+  readonly disableAssetPublishing?: boolean;
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/cloudformation-deployments.ts]: Yes please! Negative arguments break my brains.

File: packages/aws-cdk/lib/build.ts (hunk 3 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/build.ts:7]: `--concurrency` is being re-rolled, but I'm honestly not quite sure if building in parallel makes sense. The logs will probably intersperse in unreadable ways if we do that.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**